torch/csrc/jit/codegen/cuda/kernel_db/kernel_db.cpp

```diff
@@ -0,0 +1,253 @@
+  {
+    FUSER_PERF_SCOPE("KernelDb::open::read_db_txt_file");
+
+    kernel_db_txt_file_ = kernel_db_path_ / kernel_db_file;
+    if (fs::is_regular_file(kernel_db_txt_file_)) {
+      std::ifstream in_file(kernel_db_txt_file_.c_str(), std::ios::in);
+      if (in_file) {
+        bool matched_header = false;
+        bool read_db_file = true;
+        std::regex db_line_regex(
+            "^([\\w-]+),([\\w -\\=]+),([\\w-\\/]+\\.cu),([\\w-\\/]+\\.cubin)$");
+        for (std::string line; std::getline(in_file, line);) {
+          if (!matched_header) {
+            if (line.compare(header) == 0) {
+              matched_header = true;
+            } else {
+              // Header is corrupted or badly formed
+              TORCH_WARN(
+                  "Kernel DB: CSV file header is corrupted or badly formed - Resetting!: ",
+                  line);
+              read_db_file = false;
+              break;
+            }
+          } else {
+            std::smatch db_line_match;
+            if (std::regex_match(line, db_line_match, db_line_regex)) {
+              if (db_line_match.size() == 5) {
+                KernelDbEntry temp{
+                    db_line_match[1],
+                    db_line_match[2],
+                    db_line_match[3],
+                    db_line_match[4]};
+
+                fs::path code_path = kernel_db_path_ / temp.kernel_code_file;
+                std::string code;
+                if (copy_from_text_file(code_path.string(), code)) {
+                  kernel_map_[code] = temp;
+                } else {
+                  TORCH_WARN(
+                      "Kernel DB: Unable to copy cuda file: ",
+                      code_path.string());
+                }
+              }
+            } else {
+              TORCH_WARN("Kernel DB: CSV line Doesn't match: ", line);
+            }
+          }
+        }
+        if (read_db_file) {
+          return true;
+        }
+      }
+    }
+  }
+
+  // If reading of the CSV file was successful, the rest of this method
+  // is skipped
+
+  // Remove all files from directory if valid db txt file was not found
+  for (auto& path : fs::directory_iterator(kernel_db_path_)) {
+    fs::remove_all(path);
+  }
+
+  // Create an empty db csv file
+  {
+    FUSER_PERF_SCOPE("KernelDb::open::create_db_txt_file");
+
+    if (copy_to_text_file(kernel_db_txt_file_, header + "\n")) {
+      return true;
+    }
+  }
+  return false;
+}
+
+KernelDb& KernelDb::get() {
+  const std::string kernel_db_dir = "nvfuser_kernel_db";
+  const std::string kernel_db_file = "db.csv";
+
+  return get(
+      kernel_db_dir,
+      kernel_db_file,
+      true,
+      !isOptionEnabled(EnableOption::KernelDb),
+      false);
+}
+
+KernelDb& KernelDb::get(
+    const std::string& kernel_db_dir,
+    const std::string& kernel_db_file,
+    bool use_temp_dir,
+    bool disabled,
+    bool reset) {
+  std::lock_guard<std::mutex> guard(kernel_db_lock);
```

Comment:
Do we need something more aggressive than this like a file lock?

Reply:
One fun fact is file locks need to be specific for Linux and Windows.  Windows has an exclusive access API whereas Linux usually achieves this through placing a file in a specific directory.  C++ does not currently provide a standardized API.